### PR TITLE
[cli] Add dev address mode for compiler

### PR DESCRIPTION
**Pull request by gregnazario**
### Description
This allows --dev to be added to use the [dev-addresses] as well as updates that really confusing error message to match

### Test Plan
CI, and will test manually in a bit

https://github.com/aptos-labs/aptos-core/issues/8707

### CODE REVIEW

File: aptos-move/framework/src/built_package.rs [hunk 1]
```diff
@@ -39,6 +39,9 @@ pub const UPGRADE_POLICY_CUSTOM_FIELD: &str = "upgrade_policy";
 /// Represents a set of options for building artifacts from Move.
 #[derive(Debug, Clone, Parser, Serialize, Deserialize)]
 pub struct BuildOptions {
+    /// Dev mode to use [dev-addresses]
```
Comment (banool) [inline aptos-move/framework/src/built_package.rs:42]: This doesn't really explain what this flag does, can you explain this in way more detail? My assumption is when you set this, it makes the compiler take into account dev-addresses?

**Review by banool — COMMENTED** [review]
I don't really get the point of this flag, could you explain? Like aren't dev-addresses only for useful for tests?

Also we should get in the practice of writing tests for new features like this, could you please add something to `crates/aptos/e2e`?

**Review by movekevin — COMMENTED** [review]
This is awesome! Thanks for fixing this. Can we test this on the unit tests in https://github.com/aptos-labs/aptos-core/blob/resource_accounts/aptos-move/move-examples/tests/move_unit_tests.rs#L195 as well ?

File: aptos-move/framework/src/built_package.rs (hunk 1 above)
Comment (gregnazario) [reply aptos-move/framework/src/built_package.rs:42]: `dev-addresses` are never used unless you set this flag, and it's always been set to false on the compiler.

**Comment by gregnazario** [comment]
> This is awesome! Thanks for fixing this. Can we test this on the unit tests in https://github.com/aptos-labs/aptos-core/blob/resource_accounts/aptos-move/move-examples/tests/move_unit_tests.rs#L195 as well ?

Yeah, I was hoping adding this should let us use these as well

**Comment by wrwg** [comment]
Is this also addressing  #5051?

**Comment by gregnazario** [comment]
> Is this also addressing #5051?

Yeah, I believe this should also address this

**Review by wrwg — APPROVED** [review]

**Review by banool — DISMISSED** [review]
Please update the comment for the flag to explain how it works with and without the flag.

**Review by banool — CHANGES_REQUESTED** [review]
Just to be explicit that I'm waiting for the comment to be updated hahah.

**Comment by gregnazario** [comment]
> Just to be explicit that I'm waiting for the comment to be updated hahah.

I did update the comment

**Review by banool — APPROVED** [review]

**Comment by brmataptos** [comment]
Seems to fix https://github.com/aptos-labs/aptos-core/issues/5051

**Pull request merged**